src/css3.rs

```diff
@@ -1,120 +1,959 @@
+    }
+}
+
+
+
+pub struct CssAst {
+    tree: Node,
+}
+
+impl CssAst {
+    pub fn new() -> CssAst {
+        CssAst {
+            tree: Node {
+                name: "".to_string(),
+                attributes: HashMap::new(),
+                children: Vec::new(),
+            },
+        }
+    }
+
+    pub fn compile(&mut self, stylesheet: Stylesheet) {
+        let mut hm = HashMap::new();
+        hm.insert("location".to_string(), stylesheet.location.unwrap());
+
+        self.tree = Node {
+            name: "stylesheet".to_string(),
+            attributes: hm,
+            children: Vec::new(),
+        };
+
+        for rule in stylesheet.rules {
+            let node = match rule {
+                Rule::NormRule(norm_rule) => {
+                    self.compile_norm_rule(norm_rule)
+                }
+                Rule::AtRule(at_rule) => {
+                    self.compile_at_rule(at_rule)
+                }
+                Rule::QualifiedRule(qrule) => {
+                    self.compile_qualified_rule(qrule)
+                }
+            };
+
+            self.tree.children.push(node);
+        }
+    }
+
+    pub fn tree(&self) -> &Node {
+        &self.tree
+    }
+
+    pub fn display_tree(&self) {
+        self.display_tree_node(&self.tree, 0);
+    }
+
+    fn display_tree_node(&self, node: &Node, level: usize) {
+        println!("{}{}: {:?}", " ".repeat(level), node.name, node.attributes);
+        for child in &node.children {
+            self.display_tree_node(&child, level + 1);
+        }
+    }
+
+    fn compile_norm_rule(&mut self, _norm_rule: NormRule) -> Node {
+        let node = Node {
+            name: "norm-rule".to_string(),
+            attributes: HashMap::new(),
+            children: Vec::new(),
+        };
+
+        node
+
+    }
+
+    fn compile_at_rule(&mut self, _at_rule: AtRule) -> Node {
+        let node = Node {
+            name: "at-rule".to_string(),
+            attributes: HashMap::new(),
+            children: Vec::new(),
+        };
+
+        node
+    }
+
+    fn compile_qualified_rule(&mut self, qrule: QualifiedRule) -> Node {
+        let node = Node {
+            name: "qualified-rule".to_string(),
+            attributes: HashMap::new(),
+            children: Vec::new(),
+        };
+
+        let ts = TokenStream::new_from_componentvalues(qrule.prelude);
+        self.parse_prelude(ts.tokens);
+
+        // let mut p = CSS3ParserTng::new(&mut binding);
+        // let result = p.parse_commaseparated_list_of_component_values();
+        //
+        // println!("-----------------------------");
+        // println!("result: {:?}", result);
+        // println!("-----------------------------");
+
+        node
+    }
+
+    fn parse_prelude(&mut self, input: Vec<Token>) {
+        // let mut selectors = Vec::new();
+        let mut cur_selector = Vec::new();
+
+        let mut idx = 0;
+        loop {
+            let t = input[idx].clone();
+            idx += 1;
```

Comment:
Are you somewhere exiting this loop? Because else it can (and probably will) panic because of an overflow. Maybe I miss something. (and please add the semicolons after the `push` and maybe change the double quotes to single ones in line 880)